File: src/2015/05/day.05.ts
import fs from "fs";

function parseInput() {
  return fs
    .readFileSync("src/inputs/2015/05/day.05.input.txt")
    .toString()
    .split("\n")
    .filter((x) => x);
}

export function part1() {
  return getNiceStringCount(isNiceString1);
}

export function part2() {
  return getNiceStringCount(isNiceString2);
}

function getNiceStringCount(isNice: (string: string) => boolean) {
  const strings = parseInput();

  return strings.filter(isNice).length;
}

function isNiceString1(string: string) {
  // NOTE - Must contain at least 3 vowels.

  if ((string.match(/[aeiou]/g)?.length ?? 0) < 3) {
    return false;
  }

  // NOTE - Must contain a double letter.

  if (!string.match(/(.)\1/)) {
    return false;
  }

  // NOTE - Must not contain the strings ab, cd, pq, or xy.

  if (string.match(/ab|cd|pq|xy/)) {
    return false;
  }

  return true;
}

function isNiceString2(string: string) {
  // NOTE - Must contain a pair of letters appearing twice without overlapping.

  if (!string.match(/(.{2}).*\1/)) {
    return false;
  }

  // NOTE - Must contain a letter that repeats with exactly one letter between them.

  if (!string.match(/(.).\1/)) {
    return false;
  }

  return true;
}

File: src/2015/05/day.05.py
from re import search
from timeit import default_timer as timer


def parse_input():
    with open("src/inputs/2015/05/day.05.input.txt", "r") as file:
        return file.readlines()


def part1():
    strings = parse_input()
    return len([x for x in strings if is_nice_v1(x)])


def part2():
    strings = parse_input()
    return len([x for x in strings if is_nice_v2(x)])


def is_nice_v1(value: str) -> bool:
    # It does not contain the strings ab, cd, pq, or xy.
    for substring in ["ab", "cd", "pq", "xy"]:
        if value.find(substring) >= 0:
            return False

    # It contains at least three vowels.
    property_1 = len([x for x in value if x in "aeiou"]) >= 3

    # It contains at least one letter that appears twice in a row.
    property_2 = search(r"([a-z])\1{1,}", value) is not None

    return property_1 and property_2


def is_nice_v2(value: str) -> bool:
    # It contains a pair of any two letters that appears at least twice in the
    # string without overlapping.
    property_1 = search(r"([a-z]{2}).*\1", value) is not None

    # It contains at least one letter which repeats with exactly one letter
    # between them.
    property_2 = search(r"([a-z]).\1", value) is not None

    return property_1 and property_2


def timed(fn):
    start = timer()
    result = fn()

    return (timer() - start, result)


(p1_elapsed, p1_answer) = timed(part1)
print(f"Part 1: {p1_answer} ({p1_elapsed:.3f}s)")

(p2_elapsed, p2_answer) = timed(part2)
print(f"Part 2: {p2_answer} ({p2_elapsed:.3f}s)")
